Additions to the contributing guide (#8171)
MichaIng/DietPi: CONTRIBUTING.md

DietPi Project Contribution

> =============================================

Reviewer: Native Markdown heading syntax, starting with H1:
```suggestion
# DietPi Project Contribution
```

#### Are you able to?:
- Provide feedback and/or test areas of DietPi, to improve the user experience [Contributing notes](https://github.com/MichaIng/DietPi/blob/master/CONTRIBUTING.md)
- Add another [software title](https://github.com/MichaIng/DietPi/wiki/How-to-add-a-new-software-title)
- Report bugs using the [template](https://github.com/MichaIng/DietPi/issues/new?template=bug_report.md)
- Compile software for our [supported platforms](https://dietpi.com/docs/hardware/)

> - Improve/add more features to the DietPi [website](https://github.com/MichaIng/DietPi-Website) or [docs](https://github.com/MichaIng/DietPi-Docs)

Reviewer: This looks weird as a heading, maybe fine to have it as sort of introduction below the primary heading, like it was in the README. And it seems the contributing notes link was a copy&paste mistake, as it look off here?
```suggestion
Are you able to:

- Provide feedback and/or test areas of DietPi, to improve the user experience?
- Report bugs using the [this template](https://github.com/MichaIng/DietPi/issues/new?template=bug_report.md)?
- Improve or add features to DietPi, our [website](https://github.com/MichaIng/DietPi-Website) or [documentation](https://github.com/MichaIng/DietPi-Docs)?
- Add new [software titles](https://github.com/MichaIng/DietPi/wiki/How-to-add-a-new-software-title)?
- Implement support for new single-board computers?
```
I also reordered and complemented the list. "Compile software for our supported platforms" is a weird point, as we do not compile per-SBC software, and generally compile own software packages only in [rare cases](https://github.com/MichaIng/DietPi/tree/dev/.build/software). Instead, adding support for new SBCs is something which can be done often quite easily. A dedicated wiki page or section here about this makes sense, but lets have this point as typical way to contribute in this list.

If so, let us know! We are always looking for talented people who believe in the DietPi project, and, wish to contribute in any way you can.

> #### Developers:

Reviewer: H2 after H1, remove trailing `:`, and have an empty line below every heading (otherwise common Markdown linters complain):
```suggestion
## Developers

```

- Git coders: use the active development branch: [dev](https://github.com/MichaIng/DietPi/tree/dev).
- See [here](https://github.com/MichaIng/DietPi/blob/dev/BRANCH_SYSTEM.md) for repository branch guidance.
- Read below for developer-focused guidance and quick reference.

#### Other Links:
- [DietPi forum](https://dietpi.com/forum/)
- [Contribute page](https://dietpi.com/contribute.html)

> - [Security policy](https://github.com/MichaIng/DietPi/security/policy)

Reviewer: Some are redundant with README etc, but if we have relevant links here already, let's have all of them:
```suggestion
## Other Links

- [DietPi website](https://dietpi.com/)
- [DietPi documentation](https://dietpi.com/docs/)
- [DietPi forum](https://dietpi.com/forum/)
- [DietPi blog](https://dietpi.com/blog/)
- [Contribute page](https://dietpi.com/contribute.html)
- [Security policy](https://github.com/MichaIng/DietPi/security/policy)
```

#### Contact:
- micha@dietpi.com

> - GitHub: [MichaIng/DietPi](https://github.com/MichaIng/DietPi)

Reviewer: ```suggestion
## Contact

- <micha@dietpi.com>
- GitHub: [MichaIng/DietPi](https://github.com/MichaIng/DietPi)
```
Or we merge this with "Other Links", to something like "Other entry points", "Engage", "Contact and community" or such? At least the GitHub and forum links both fit below both headings, and for the email address only, a dedicated heading is probably overkill?

## DietPi Script Development

The below guide should help contributors add or modify DietPi scripts under
`/boot/dietpi` with minimal friction. It focuses on safe extension points,
shared helpers, persistence conventions, and reviewer-friendly test plans.

Branch workflow

> ---------------

Reviewer: ```suggestion
### Branch workflow

```

> DietPi separates stable releases, pre-release testing and active development:

Reviewer: ```suggestion
The DietPi Git repository is separated into three branches:
```

- `master`: stable release branch (production-ready).
- `beta`: public [pre-release testing branch](https://github.com/MichaIng/DietPi/blob/dev/BRANCH_SYSTEM.md).
- `dev`: active [development branch](https://github.com/MichaIng/DietPi/tree/dev) — implement new features and fixes here.

Guidance:

- Target `dev` for code contributions and open PRs against it unless asked otherwise.

> - Switching branches on-device:

Reviewer: ```suggestion
- Switching branches on your DietPi device:
```

>   - Fresh image: edit `/boot/dietpi.txt` and change `DEV_GITBRANCH=` then reboot.

Reviewer: ```suggestion
  - Fresh image: Before first boot, edit `dietpi.txt` on the FAT partition of the flashed image, and change `DEV_GITBRANCH=`. Adjust `DEV_GITOWNER=` accordingly, if you want to switch to your fork of DietPi.
```

>   - Existing install: run `dietpi-backup` first, then use `G_DEV_BRANCH beta` or `G_DEV_BRANCH dev` to switch.

Reviewer: ```suggestion
  - Existing install: run `dietpi-backup` first, then use `G_DEV_BRANCH beta` or `G_DEV_BRANCH dev` to quickly to the respective branch, invoking `dietpi-update`.
```

> - Warnings: `dev` and `beta` can be unstable; avoid using them on critical systems. Always include a Test Plan and restore steps in PRs when testing pre-release branches.

Reviewer: ```suggestion
- Warning: `beta` and especially `dev` can be unstable; avoid using them on critical production systems. Have a `dietpi-backup`, or do tests on a dedicated testing system.
```

Core concepts

> -------------

Reviewer: ```suggestion
### Core concepts

```

> - Globals: scripts source `func/dietpi-globals` for shared helpers and

Reviewer: Let's use the longest part of the path that matches between Git repo and DietPi system:
```suggestion
- Globals: scripts source `dietpi/func/dietpi-globals` for shared helpers and
```

  environment setup (`G_INIT`, `G_EXIT`, color vars, etc.). Read this file
  first to understand helper semantics and cancel/error behavior.
- UI: prefer `G_WHIP_*` dialog helpers for menus, input and confirmations to
  maintain a consistent user experience across scripts.
- Persistence: use `FP_SAVEFILE` / `FP_SETTINGS` to write small shell-sourcable
  files. Persist arrays as indexed assignments (e.g. `aENABLED[3]=1`) to

>   preserve compatibility across edits.

Reviewer: This point is unclear to me: `FP_SAVEFILE`/`FP_SETTINGS` are just random variables used in some scripts for longer config file paths (which are indeed often shell scripts to source variables from). Indexed arrays are similarly often used. However, whether a dedicated config file is reasonable for a new script, which format it has, whether its path is stored in a variable, and in case which name it has, is nothing I'd want to suggest in general. It depends, and indexed arrays are actually something I want to use less often. E.g. associative arrays can use self-explaining textual indexes, so that one does not need to lookup the mapping every time when browsing through the script. Or you mean to keep the old indexes intact when adding new ones? That probably/hopefully does not need to be mentioned 😃.
```suggestion
```

Author: I see future issues (until indexes are ALL textual) with dietpi-banner in particular, since the options are integer indexed. For example, when I was adding my functions, they had to go at the end of aENABLED so that existing settings would not get clobbered. Maybe need a separate workthrough to import and translate old indexes?

Author: I changed the Menu Extension section and clarified this section

### Core concepts

- Globals: scripts source `dietpi/func/dietpi-globals` for shared helpers and
  environment setup (`G_INIT`, `G_EXIT`, color vars, etc.). Read this file
  first to understand helper semantics and cancel/error behavior.
- UI: prefer `G_WHIP_*` dialog helpers for menus, input and confirmations to
  maintain a consistent user experience across scripts.
- Error-handling: use `G_EXEC` to wrap any command call, so DietPi's error
  handler and consistent console output apply. Validate root permissions and
  write access using `G_CHECK_ROOT_USER` / `G_CHECK_ROOTFS_RW` where required.
- Persistence: Write arrays or variables into a preference file `/boot/dietpi/.<prog_settings>`.
  Persist arrays as indexed assignments (e.g. `aARRAY[index]=1`). Convert ESC bytes to `\e`
  when saving text if needed. Load the preferences with `. "/boot/dietpi/.<prog_settings>"`

### Menu extension pattern (safe, minimal)

1. Add handler: implement `Menu_<Name>()` to present inputs (use `G_WHIP_*`),
   validate, and update in-memory variables (e.g. `aENABLED[index]`).
2. Register option: add the menu label into `Menu_Main()` (scripts use a
   case-switch dispatch). Remember to update `MENU_LASTITEM_*` indices if used.
3. Persist: Write arrays or variables into a preference file (see the 'Core concepts' section).
4. Test: include interactive steps in your PR Test Plan (open menu, toggle,
   verify `cat /boot/dietpi/.<prog_settings>`).

- Safe system changes: use `G_EXEC` for any commands that modify the system so
  DietPi's error handling and reporting apply. Always validate root/write

>   access using `G_CHECK_ROOT_USER` / `G_CHECK_ROOTFS_RW` where required.

Reviewer: ```suggestion
- Error-handling: use `G_EXEC` to wrap any command call, so DietPi's error
  handler and consistent console output apply. Validate root permissions and
  write access using `G_CHECK_ROOT_USER` / `G_CHECK_ROOTFS_RW` where required.
```

Helper functions (G_ helpers)

> -----------------------------

Reviewer: ```suggestion
### Helper functions (G_ helpers)

```

> DietPi provides many `G_` prefixed helpers in `func/dietpi-globals`. Below

Reviewer: ```suggestion
DietPi provides many `G_` prefixed helpers in `dietpi/func/dietpi-globals`. Below
```

are the most useful ones for contributors and how to use them safely.

- `G_INIT` — initialize script runtime, set traps, and handle concurrent

>   execution checks. Call early after sourcing globals. It sets up `G_EXIT`.

Reviewer: ```suggestion
- `G_INIT` — initialize script runtime, sets up the working directory, exit
   traps, consistent locale for parsing external command outputs, and handles
   concurrent execution checks. Call early after sourcing `dietpi-globals`.
```
We could just put the typical script header here as example:
```sh
. /boot/dietpi/func/dietpi-globals
readonly G_PROGRAM_NAME='DietPi-Config'
G_CHECK_ROOT_USER "$@" # if the script requires root permissions
G_CHECK_ROOTFS_RW # if the script requires write access
G_INIT
```
Could be mentioned that `G_CHECK_ROOT_USER "$@"` also attempts to re-execute the script with `sudo`, if needed. This is the reason CLI args are passed through.

Author: I clarified the G_ section and added a sample script header

- `G_EXIT` — cleanup/exit handler registered by `G_INIT`. Avoid overriding
  unless you re-register a compatible trap; use `G_EXIT` to ensure proper
  teardown on SIGINT/EXIT.

> 

Reviewer: `G_EXIT` does not need to be called manually. `G_INIT` sets it up as EXIT trap, i.e. it is called automatically.
```suggestion
```
What could make sense is `G_EXIT_CUSTOM`, which can be defined as function, and is then called by `G_EXIT`. Could be used for additional script-specific cleanup steps.

- `G_EXEC` — robust command executor with built-in retries and an interactive
  error handler. Use instead of direct `rm`/`systemctl` in scripts so
  failures are presented to the user and logged consistently. Optional
  env vars: `G_EXEC_DESC`, `G_EXEC_RETRIES`, `G_EXEC_OUTPUT`.

- `G_CONFIG_INJECT` — targeted config-file editing helper. Use to atomically
  replace or add config lines using predictable patterns rather than ad-hoc

>   `sed` calls.

Reviewer: Additionally, if the config line is present in commented form, `G_CONFIG_INJECT` uncomments it. That it merges an `if ... elif ... else ... fi` complex to apply configs in a clean way is the reason I added it once:
```suggestion
- `G_CONFIG_INJECT` — targeted config-file editing helper. Use to atomically
  replace, uncomment, or add config lines using predictable patterns rather
  than ad-hoc `sed` calls.
```

- `G_WHIP_*` family — dialog and UI helpers (`G_WHIP_MENU`,

>   `G_WHIP_CHECKLIST`, `G_WHIP_INPUTBOX`, `G_WHIP_YESNO`, `G_WHIP_VIEWFILE`).

Reviewer: ```suggestion
- `G_WHIP_*` family — dialog and UI helpers: (`G_WHIP_MSG`, `G_WHIP_YESNO`,
  `G_WHIP_MENU`, `G_WHIP_CHECKLIST`, `G_WHIP_INPUTBOX`, `G_WHIP_PASSWORD`,
  `G_WHIP_VIEWFILE`).
```

  Prefer these for user interaction to maintain consistent UX and behavior.

G_WHIP specifics

> ----------------

Reviewer: ```suggestion
#### G_WHIP specifics

```

These details are commonly needed when implementing menus and input boxes.

- `G_WHIP_INPUTBOX`:
  - Use `G_WHIP_INPUTBOX_REGEX` to provide a validation regex and

>     `G_WHIP_INPUTBOX_REGEX_TEXT` to describe allowed input (human-friendly).

Reviewer: ```suggestion
  - Use `G_WHIP_INPUTBOX_REGEX` to provide an optional validation regex and
    `G_WHIP_INPUTBOX_REGEX_TEXT` to describe allowed input (human-friendly).
    By default, any non-empty input is allowed.
```

  - `G_WHIP_DEFAULT_ITEM` pre-fills the input field. The helper loops until
    input matches the regex or the user cancels (`|| return`).
  - On success the entered value is returned in `$G_WHIP_RETURNED_VALUE`.

- `G_WHIP_YESNO`:
  - Presents a Yes/No dialog. Exit status `0` indicates Yes; non-zero is
    No or cancel. Some helper versions may also set `$G_WHIP_RETURNED_VALUE`.
  - Use for confirmations before destructive actions. Combine with

>     `G_EXEC` for safe command execution on confirmation.

Reviewer: This did never set `G_WHIP_RETURNED_VALUE`, just a return code:
```suggestion
- `G_WHIP_YESNO`:
  - Presents a Yes/No dialog. Exit status `0` indicates Yes/Ok; non-zero is
    No/Cancel.
  - Use for confirmations before destructive actions. Combine with
    `G_EXEC` for safe command execution on confirmation.
```

> - `G_WHIP_DEFAULT_ITEM`:

Reviewer: A bad design choice in DietPi was to not differentiate well between `G_` functions and `G_` variables by name or any such. Especially in such list here, where helper functions and variables consumed or set by those are mixed, this can be confusing. I suggest we use `$G_FOO` `$` prefix for variables, and `G_BAR()` `()` suffix for functions consistently across this document. The below one just as an example/starting point:
```suggestion
- `$G_WHIP_DEFAULT_ITEM`:
```

Author: This seems like a larger re-factor, involving renames, probably worth it's own PR?

Reviewer: I did not mean to actually rename the variables (where `$` is not allowed anyway), but just to separate them better from functions in this document. Since variables are called with `$` prefix, this should be a natural intuitive way. Adding `()` as suffix to functions (as this is how functions need to be declared in scripts), is probably not needed anymore. So the way it is now (after my last commits), `G_*` functions and variables both can be copied from the document and pasted into a shell as is to call them (while previously functions were in their call form, but variables in their declaration/assignment form).

  - Controls the pre-selected menu item or prefilled input box value.
  - When using `G_WHIP_MENU`, set it to a label matching one of the menu
    entries to pre-select that entry (exact match is used).

- `G_WHIP_SIZE_X_MAX`:
  - Optional integer to limit dialog width (chars). Useful for very long

>     content to keep dialogs within readable proportions on narrow terminals.

Reviewer: ```suggestion
  - Optional integer to limit dialog width (chars). Useful for dialogs with
    very few content or checklists, where too wide dialog boxes may look weird.
    By default, all `G_WHIP` dialogs are max 120 characters wide, capped by the
    terminal width.
```

  - Set it before calling a `G_WHIP_*` helper; the helper respects it when
    calculating `WHIP_SIZE_X`.

- `G_CHECK_ROOT_USER`, `G_CHECK_ROOTFS_RW` — validate that the script runs
  with necessary privileges and writable rootfs before performing writes.

- `G_GET_NET`, `G_GET_WAN_IP` — network helpers that return standardized

>   values; use `-q` to request raw values suitable for scripting.

Reviewer: ```suggestion
- `G_GET_NET`, `G_GET_WAN_IP` — network helpers that return standardized
  values; use `-q` to hide error messages.
```

- `G_TRUNCATE_MID` — utility to shorten long strings with `...` in the
  middle; useful for CLI-friendly output.

> 

Reviewer: I see you add this in #8172. Since I won't be able to review that one before DietPi v10.5 release, let's remove it here, and add in #8172 after merge.
```suggestion
```

- `G_DIETPI-NOTIFY` / `G_BUG_REPORT` — helpers to generate formatted bug
  reports and diagnostics. Use when capturing logs for PRs / issues.

Usage hints:
- Read `func/dietpi-globals` when adding behavior that interacts with the
  user, modifies files, or runs external commands — it documents optional

>   environment variables and exit/cancel semantics for each helper.

Reviewer: ```suggestion
- Read `dietpi/func/dietpi-globals` when adding behavior that interacts
  with the user, modifies files, or runs external commands — it documents
  optional environment variables and exit/cancel semantics for each helper.
```

- Prefer the `G_` helpers over ad-hoc implementations to keep error handling
  consistent and reduce reviewer friction.

Menu extension pattern (safe, minimal)

> --------------------------------------

Reviewer: ```suggestion
### Menu extension pattern (safe, minimal)

```

1. Add handler: implement `Menu_<Name>()` to present inputs (use `G_WHIP_*`),
   validate, and update in-memory variables (e.g. `aENABLED[index]`).
2. Register option: add the menu label into `Menu_Main()` (scripts use a
   case-switch dispatch). Remember to update `MENU_LASTITEM_*` indices if used.
3. Persist: call `Save` or `Write_Settings_File()` to write to `FP_SAVEFILE`.
   Persist arrays as `aENABLED[index]=...` lines; convert ESC bytes to `\e`
   when saving color slots if needed.
4. Test: include interactive steps in your PR Test Plan (open menu, toggle,
   verify `cat /boot/dietpi/.<prog_settings>`).

Practical snippets (quick reference)

> -----------------------------------

Reviewer: ```suggestion
### Practical snippets (quick reference)

```

Below are minimal, copy-paste-ready examples that follow DietPi conventions.

- `G_WHIP_MENU` (single choice):
```
G_WHIP_MENU_ARRAY=( 'Start' 'Start the service' 'Stop' 'Stop the service' )
G_WHIP_DEFAULT_ITEM='Start'
G_WHIP_MENU 'Select action:' || return
case $G_WHIP_RETURNED_VALUE in
  Start) echo 'Starting...';;
  Stop) echo 'Stopping...';;
esac

> ```

Reviewer: That way it is indented to match the list:
````suggestion
    ```
    G_WHIP_MENU_ARRAY=( 'Start' 'Start the service' 'Stop' 'Stop the service' )
    G_WHIP_DEFAULT_ITEM='Start'
    G_WHIP_MENU 'Select action:' || return
    case $G_WHIP_RETURNED_VALUE in
      Start) echo 'Starting...';;
      Stop) echo 'Stopping...';;
    esac
    ```
````

- `G_WHIP_CHECKLIST_ARRAY` (multi-select):
```
G_WHIP_CHECKLIST_ARRAY=()
G_WHIP_CHECKLIST_ARRAY+=( '5' 'Enable Foo' "${aENABLED[5]:=0}" )
G_WHIP_CHECKLIST_ARRAY+=( '6' 'Enable Bar' "${aENABLED[6]:=0}" )
G_WHIP_CHECKLIST 'Choose features to enable:' || return
for i in $G_WHIP_RETURNED_VALUE; do aENABLED[$i]=1; done
Save > "$FP_SAVEFILE"
```
- `G_WHIP_INPUTBOX` (validated input):
```
G_WHIP_INPUTBOX_REGEX='^[0-9]+$' G_WHIP_INPUTBOX_REGEX_TEXT='a number' G_WHIP_DEFAULT_ITEM=10
G_WHIP_INPUTBOX 'Set retry count:' || return
RETRIES=$G_WHIP_RETURNED_VALUE
```
- `G_WHIP_YESNO` (confirmation):
```
if G_WHIP_YESNO 'Delete backup?'; then
  G_EXEC rm -rf "$TARGET"
fi
```
- `G_WHIP_VIEWFILE` (show a logfile):
```
log=1 G_WHIP_VIEWFILE "$FP_LOG" || return
```
- `Save()` persistence pattern (follow dietpi-banner conventions):
```
Save(){
  echo "aDESCRIPTION[10]='${aDESCRIPTION[10]}'"
  for i in "${!aENABLED[@]}"; do echo "aENABLED[$i]=${aENABLED[$i]}"; done
  for i in {0..6}; do val="${aCOLOUR[$i]}"; esc=$(printf '%s' "$val" | sed $'s/\x1b/\\e/g'); esc=${esc//\'/\\\'}; echo "aCOLOUR[$i]='$esc'"; done
}

> ```

Reviewer: ````suggestion
    ```
    G_WHIP_CHECKLIST_ARRAY=()
    G_WHIP_CHECKLIST_ARRAY+=( '5' 'Enable Foo' "${aENABLED[5]:=0}" )
    G_WHIP_CHECKLIST_ARRAY+=( '6' 'Enable Bar' "${aENABLED[6]:=0}" )
    G_WHIP_CHECKLIST 'Choose features to enable:' || return 0
    for i in $G_WHIP_RETURNED_VALUE; do aENABLED[$i]=1; done
    Save > "$FP_SAVEFILE"
    ```
- `G_WHIP_INPUTBOX` (validated input):
    ```
    G_WHIP_INPUTBOX_REGEX='^[0-9]+$' G_WHIP_INPUTBOX_REGEX_TEXT='a number' G_WHIP_DEFAULT_ITEM=10
    G_WHIP_INPUTBOX 'Set retry count:' || return 0
    RETRIES=$G_WHIP_RETURNED_VALUE
    ```
- `G_WHIP_YESNO` (confirmation):
    ```
    if G_WHIP_YESNO 'Delete backup?'
    then
    	G_EXEC rm -rf "$TARGET"
    fi
    ```
- `G_WHIP_VIEWFILE` (show a logfile):
    ```
    log=1 G_WHIP_VIEWFILE "$FP_LOG" || return 0
    ```
- `Save()` persistence pattern (follow dietpi-banner conventions):
    ```
    Save()
    {
    	echo "aDESCRIPTION[10]='${aDESCRIPTION[10]}'"
    	for i in "${!aENABLED[@]}"; do echo "aENABLED[$i]=${aENABLED[$i]}"; done
    	for i in {0..6}; do val="${aCOLOUR[$i]}"; esc=$(printf '%s' "$val" | sed $'s/\x1b/\\e/g'); esc=${esc//\'/\\\'}; echo "aCOLOUR[$i]='$esc'"; done
    }
    ```
````

Banner extension pattern (example)

> ----------------------------------

Reviewer: ```suggestion
### Banner extension pattern (example)

```

When adding banner items (e.g. `dietpi-banner`), follow this minimal pattern:

- Describe: add the label to `aDESCRIPTION[index]` and a default to
  `aENABLED[index]` during initialization.

- Output: implement `Print_<ShortName>()` or add a guarded line in `Print_Banner_raw()`:
```
(( ${aENABLED[index]} )) && echo -e "$GREEN_BULLET ${aCOLOUR[1]}${aDESCRIPTION[index]} $GREEN_SEPARATOR $(Print_<ShortName>)"

> ```

Reviewer: ````suggestion
    ```
    (( ${aENABLED[index]} )) && echo -e "$GREEN_BULLET ${aCOLOUR[1]}${aDESCRIPTION[index]} $GREEN_SEPARATOR $(Print_<ShortName>)"
    ```
````

- Persist: ensure `Save()` writes `aENABLED[index]=...` and any custom
  `aDESCRIPTION[...]` lines so the state survives restarts.

Nested / multi-page menu pattern

> --------------------------------

Reviewer: ```suggestion
### Nested / multi-page menu pattern

```

```
TARGETMENUID=0  # start at the top level main menu
while (( TARGETMENUID != -1 )); do
  case $TARGETMENUID in
    0) Menu_Main; ;;  # sets TARGETMENUID based on selection
    1) Menu_Settings; TARGETMENUID=0 ;;  # run settings page then return to main menu
    -1) break;;
  esac
done
```

AWK wrapper call (word-wrap helper)

> -----------------------------------

Reviewer: This one is for the banner in particular, while the above menu look is generic. Maybe this heading should become a sub/H4 heading within the banner heading?
```suggestion
### AWK wrapper call (word-wrap helper)

```
But also, maybe it is too specific to have it here. It is a specific feature of the banner, but does not affect contribution to/extension of the banner otherwise. If there is a script worth explain more in detail, then it would be `dietpi-software`, or some structures like `dietpi-config` frontend/menu vs `dietpi-set_software`/`dietpi-set_hardware` backend CLI scripts, or those scripts/services invoked during boot: https://github.com/MichaIng/DietPi/wiki/Booting-process

Author: I think ideally, scripts that are at higher risk/frequency of being developed, should each have a readme section at the top with documentation. Removed this section from the doc.

Reviewer: Agreed. Though probably it is better to create wiki pages for each of them, so that the size of the contribution document stays in ranges?
